include/libcyphal/presentation/subscriber.hpp

```diff
@@ -0,0 +1,145 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_PRESENTATION_SUBSCRIBER_HPP_INCLUDED
+#define LIBCYPHAL_PRESENTATION_SUBSCRIBER_HPP_INCLUDED
+
+#include "subscriber_impl.hpp"
+
+#include "libcyphal/transport/errors.hpp"
+#include "libcyphal/transport/scattered_buffer.hpp"
+#include "libcyphal/transport/types.hpp"
+
+#include <cetl/cetl.hpp>
+#include <cetl/pmr/function.hpp>
+#include <nunavut/support/serialization.hpp>
+
+#include <tuple>
+#include <utility>
+
+namespace libcyphal
+{
+namespace presentation
+{
+
+/// Internal implementation details of the Presentation layer.
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+
+// TODO: docs
+/// No Sonar cpp:S4963 "The "Rule-of-Zero" should be followed"
+/// b/c we do directly handle resources here.
+class SubscriberBase : public SubscriberImpl::CallbackNode  // NOSONAR cpp:S4963
+{
+public:
+    using Failure = transport::AnyFailure;
+
+    SubscriberBase(SubscriberBase&& other) noexcept
+        : CallbackNode{std::move(static_cast<CallbackNode&>(other))}
+        , impl_{std::exchange(other.impl_, nullptr)}
+    {
+        CETL_DEBUG_ASSERT(impl_ != nullptr, "Not supposed to move from already moved `other`.");
+        impl_->updateCallbackNode(&other, this);
+    }
+
+    SubscriberBase& operator=(SubscriberBase&& other) noexcept
+    {
+        CETL_DEBUG_ASSERT(impl_ != nullptr, "Not supposed to move to already moved `this`.");
+
+        impl_->releaseCallbackNode(*this);
+
+        static_cast<CallbackNode&>(*this) = std::move(static_cast<CallbackNode&>(other));
+
+        impl_ = std::exchange(other.impl_, nullptr);
+        impl_->updateCallbackNode(&other, this);
+
+        return *this;
+    }
+
+    SubscriberBase(const SubscriberBase& other)            = delete;
+    SubscriberBase& operator=(const SubscriberBase& other) = delete;
+
+protected:
+    ~SubscriberBase() noexcept
+    {
+        if (impl_ != nullptr)
+        {
+            impl_->releaseCallbackNode(*this);
+        }
+    }
+
+    SubscriberBase(SubscriberImpl* const impl, const Deserializer deserializer)
+        : CallbackNode{impl->now(), deserializer}
+        , impl_{impl}
+    {
+        CETL_DEBUG_ASSERT(impl != nullptr, "");
+
+        impl_->retainCallbackNode(*this);
+    }
+
+private:
+    // MARK: Data members:
+
+    SubscriberImpl* impl_;
+
+};  // SubscriberBase
+
+}  // namespace detail
+
+// TODO: docs
+template <typename Message>
+class Subscriber final : public detail::SubscriberBase
```

Comment:
Missing peek() and pop() described in the design document. Unless the user opted to use a callback, the Subscriber should act as a one-element-deep queue that can be read using the aforementioned methods. Eventually we could introduce configurable queue size if there's a need for that.

Reply:
I did it on purpose! I propose to have instead configurable adapters, which you can very easy plug into Subscriber's "on receive" callback. Idea is that...
1. Subscriber never stores/caches messages, but only delivers them via callback
2. Plugged adapter can implement whatever storage policy user wants

Reply:
Also, I thought that maybe even `transport::IXxxRxSession`-s should not have `receive` alternative... but taking into consideration that it's kinda low level api, and the fact that these rx sessions are in **single** instance per port id, I decided to keep both way to get raw messages (although they are mutually exclusive).

Reply:
Is it going to affect the user API significantly? I can't quite imagine what your proposal should look like. Ok we should talk about this tomorrow.

Reply:
Here an example of what I mean:

```
// Subscription
auto heartbeat_storage = MessageStorage<NodeHelpers::Heartbeat::Message, 1>{};
auto heartbeat_subscriber = NodeHelpers::Heartbeat::makeSubscriber(presentation);
heartbeat_subscriber->setOnReceiveCallback(std::ref(heartbeat_storage));
...
// Main loop.
//
CommonHelpers::runMainLoop(executor_, startup_time_ + run_duration_ + 500ms, [&](const auto now) {
    //
    if (auto heartbeat = heartbeat_storage.pop())
    {
        NodeHelpers::Heartbeat::print(now - startup_time_, std::get<0>(*heartbeat), std::get<1>(*heartbeat));
    }
});
```
And message storage could something like this:
```
template <typename Message, std::size_t Size>
class MessageStorage;
template <typename Message>
class MessageStorage<Message, 1>
{
public:
    using MessageAndMetadata = std::tuple<Message, MessageRxMetadata>;

    cetl::optional<MessageAndMetadata> pop() noexcept
    {
        return std::exchange(last_message_, cetl::nullopt);
    }

    cetl::optional<const MessageAndMetadata&> peek() const noexcept
    {
        return last_message_;
    }

    void operator()(const typename Subscriber<Message>::OnReceiveCallback::Arg& arg)
    {
        last_message_.emplace(arg.message, arg.metadata);
    }

private:
    cetl::optional<std::tuple<Message, MessageRxMetadata>> last_message_;

}; // MessageStorage
```

Reply:
BTW, other but similar `MessageStorage` could potentially have multi-threading protection inside if it faces multiple threads or MCU cores. In such way it will be possible to continue "run" our libcyphal on a single-threaded executor, but potentially consume messages (via peek/pop) from other threads/cores. It is just an idea/example - my main point again about concerns separation, in this case subscription vs concurrency.